lgc/patch/FragColorExport.cpp

```diff
@@ -188,45 +196,61 @@ Value *FragColorExport::handleColorExportInstructions(Value *output, unsigned hw
       extractElements(output, builder, comps);
 
+      exports[0] = exports[1] = undefFloat16x2;
       Attribute::AttrKind attribs[] = {Attribute::ReadNone};
-      comps[0] = builder.CreateNamedCall("llvm.amdgcn.cvt.pkrtz", FixedVectorType::get(builder.getHalfTy(), 2),
-                                         {comps[0], comps[1]}, attribs);
-      if (compCount > 2)
-        comps[1] = builder.CreateNamedCall("llvm.amdgcn.cvt.pkrtz", FixedVectorType::get(builder.getHalfTy(), 2),
-                                           {comps[2], comps[3]}, attribs);
+      for (unsigned idx = 0; idx < compactCompCount; ++idx) {
+        unsigned origIdx = 2 * idx;
+        unsigned compMask = (1 << origIdx) | (1 << (origIdx + 1));
+        if (compMask & channelWriteMask) {
+          exports[idx] = builder.CreateNamedCall("llvm.amdgcn.cvt.pkrtz", FixedVectorType::get(builder.getHalfTy(), 2),
+                                                 {comps[origIdx], comps[origIdx + 1]}, attribs);
+          exportMask |= compMask;
+        }
+      }
     }
-    break;
-  }
-  case EXP_FORMAT_UNORM16_ABGR:
-  case EXP_FORMAT_SNORM16_ABGR: {
-    output = convertToFloat(output, signedness, builder);
-    extractElements(output, builder, comps);
-
-    StringRef funcName =
-        expFmt == EXP_FORMAT_SNORM16_ABGR ? "llvm.amdgcn.cvt.pknorm.i16" : "llvm.amdgcn.cvt.pknorm.u16";
-
-    for (unsigned idx = 0; idx < (compCount + 1) / 2; idx++) {
-      Value *packedComps = builder.CreateNamedCall(funcName, FixedVectorType::get(builder.getInt16Ty(), 2),
-                                                   {comps[2 * idx], comps[2 * idx + 1]}, {});
-
-      comps[idx] = builder.CreateBitCast(packedComps, FixedVectorType::get(builder.getHalfTy(), 2));
+    // GFX11 removes compressed export, simply use 32bit-data export.
+    if (m_pipelineState->getTargetInfo().getGfxIpVersion().major >= 11 && !dualSourceBlendedEnable) {
+      exportMask = (1 << compactCompCount) - 1;
+      for (unsigned idx = 0; idx < compactCompCount; ++idx)
+        exports[idx] = builder.CreateBitCast(exports[idx], builder.getFloatTy());
     }
-
     break;
   }
+  case EXP_FORMAT_UNORM16_ABGR:
+  case EXP_FORMAT_SNORM16_ABGR:
   case EXP_FORMAT_UINT16_ABGR:
   case EXP_FORMAT_SINT16_ABGR: {
     assert(compCount <= 4);
-    output = convertToInt(output, signedness, builder);
-    extractElements(output, builder, comps);
 
-    StringRef funcName = expFmt == EXP_FORMAT_SINT16_ABGR ? "llvm.amdgcn.cvt.pk.i16" : "llvm.amdgcn.cvt.pk.u16";
+    StringRef funcName;
+    if (expFmt == EXP_FORMAT_SNORM16_ABGR || expFmt == EXP_FORMAT_UNORM16_ABGR) {
+      output = convertToFloat(output, signedness, builder);
+      funcName = EXP_FORMAT_SNORM16_ABGR ? "llvm.amdgcn.cvt.pknorm.i16" : "llvm.amdgcn.cvt.pknorm.u16";
+    } else {
+      output = convertToInt(output, signedness, builder);
+      funcName = expFmt == EXP_FORMAT_SINT16_ABGR ? "llvm.amdgcn.cvt.pk.i16" : "llvm.amdgcn.cvt.pk.u16";
```

Comment:
Please update this code to use intrinsic IDs instead of function names.

Reply:
Fixed.